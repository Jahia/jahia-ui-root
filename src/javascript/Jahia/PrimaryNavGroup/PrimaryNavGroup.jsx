import React, {Suspense} from 'react';
import PropTypes from 'prop-types';
import {PrimaryNavItem, PrimaryNavItemsGroup} from '@jahia/moonstone';
import {registry} from '@jahia/ui-extender';
import {useTranslation} from 'react-i18next';
import {useHistory} from 'react-router';
import {useNodeInfo} from '@jahia/data-helper';
import {useSelector} from 'react-redux';

export const PrimaryNavGroup = ({isDisplayedWhenCollapsed, target}) => {
    const {t} = useTranslation();
    const history = useHistory();
    const current = useSelector(state => ({language: state.language, site: state.site}));
    const navItems = registry.find({type: 'primary-nav-item', target});
    const requiredPermission = ['jcr:read_default'];
    const requiredPaths = [];
    navItems.filter(navItem => navItem.requiredPermission !== undefined).forEach(navItem => {
        if (!requiredPermission.includes(navItem.requiredPermission)) {
            requiredPermission.push(navItem.requiredPermission);
        }

        if (navItem.requiredPermissionPath !== undefined && !requiredPaths.includes(navItem.requiredPermissionPath)) {
            requiredPaths.push(navItem.requiredPermissionPath);
        }
    });
    if (requiredPaths.length === 0) {
        requiredPaths.push('/sites/' + current.site);
    }

    const permissions = useNodeInfo({paths: requiredPaths, language: 'en'},
        {getPermissions: requiredPermission});
    if (permissions.loading === true || permissions.nodes === null) {
        return null;
    }

    console.log(permissions);

    return (
        <Suspense fallback="loading...">
            <PrimaryNavItemsGroup isDisplayedWhenCollapsed={isDisplayedWhenCollapsed}>
                {navItems.filter(navItem => navItem.requiredPermission === undefined ||
                    permissions.nodes.find(node => {
                        if (navItem.requiredPermissionPath !== undefined) {
                            return node.path === navItem.requiredPermissionPath;
                        }

                        return node.path === '/sites/' + current.site;
                    })[navItem.requiredPermission]).map(item => {
                    if (item.render) {
                        return React.cloneElement(item.render(), {key: item.key});
                    }

                    return (
                        <PrimaryNavItem key={item.key}
                                        role={`${item.key}-menu-item`}
                                        isSelected={history.location.pathname.startsWith(item.path)}
                                        icon={item.icon}
                                        label={t(item.label)}
                                        onClick={() => history.push(item.path)}/>
                    );
                })}
            </PrimaryNavItemsGroup>
        </Suspense>
    );
};

PrimaryNavGroup.defaultProps = {
    isDisplayedWhenCollapsed: true
};

PrimaryNavGroup.propTypes = {
    isDisplayedWhenCollapsed: PropTypes.bool,
    target: PropTypes.string.isRequired
};

export default PrimaryNavGroup;
