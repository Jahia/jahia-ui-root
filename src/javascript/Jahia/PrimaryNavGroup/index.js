import PrimaryNavGroup from './PrimaryNavGroup';

export default PrimaryNavGroup;
