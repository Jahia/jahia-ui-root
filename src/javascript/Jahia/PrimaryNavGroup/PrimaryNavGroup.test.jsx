import React from 'react';
import {render} from '@testing-library/react';
import PrimaryNavGroup from './PrimaryNavGroup';

jest.mock('react-router', () => {
    return {
        useHistory: jest.fn(() => {
        })
    };
});

jest.mock('react-redux', () => {
    return {
        useSelector: jest.fn(() => ({
            site: 'systemsite',
            language: 'en'
        }))
    };
});

jest.mock('@jahia/data-helper', () => {
    return {
        useNodeInfo: jest.fn(() => ([{
            data: {
                jcr: {
                    nodesByPath: [{
                        name: 'systemsite',
                        uuid: 'fd148dc0-1a97-4eb0-8ec6-fbeee9ddbdff',
                        workspace: 'EDIT',
                        path: '/sites/systemsite',
                        __typename: 'JCRSite',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }], __typename: 'JCRQuery'
                }
            }
        }, {
            data: {
                jcr: {
                    nodesByPath: [{
                        name: 'systemsite',
                        uuid: 'fd148dc0-1a97-4eb0-8ec6-fbeee9ddbdff',
                        workspace: 'EDIT',
                        path: '/sites/systemsite',
                        __typename: 'JCRSite',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }, {
                        name: '',
                        uuid: 'cafebabe-cafe-babe-cafe-babecafebabe',
                        workspace: 'EDIT',
                        path: '/',
                        __typename: 'GenericJCRNode',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }], __typename: 'JCRQuery'
                }
            }
        }, {
            data: {
                jcr: {
                    nodesByPath: [{
                        name: 'systemsite',
                        uuid: 'fd148dc0-1a97-4eb0-8ec6-fbeee9ddbdff',
                        workspace: 'EDIT',
                        path: '/sites/systemsite',
                        __typename: 'JCRSite',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }], __typename: 'JCRQuery'
                }
            }
        }, {
            data: {
                jcr: {
                    nodesByPath: [{
                        name: 'systemsite',
                        uuid: 'fd148dc0-1a97-4eb0-8ec6-fbeee9ddbdff',
                        workspace: 'EDIT',
                        path: '/sites/systemsite',
                        __typename: 'JCRSite',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }], __typename: 'JCRQuery'
                }
            }
        }, {
            data: {
                jcr: {
                    nodesByPath: [{
                        name: 'systemsite',
                        uuid: 'fd148dc0-1a97-4eb0-8ec6-fbeee9ddbdff',
                        workspace: 'EDIT',
                        path: '/sites/systemsite',
                        __typename: 'JCRSite',
                        // eslint-disable-next-line camelcase
                        permission_encoded_amNyOnJlYWRfZGVmYXVsdA: true
                    }], __typename: 'JCRQuery'
                }
            }
        }]))
    };
});

describe('primary nav group', () => {
    test('collapsed and empty nav bar', () => {
        const {asFragment} = render(
            <PrimaryNavGroup isDisplayedWhenCollapsed target="none"/>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
