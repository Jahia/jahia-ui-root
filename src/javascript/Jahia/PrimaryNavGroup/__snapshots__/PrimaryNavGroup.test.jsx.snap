// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`primary nav group collapsed and empty nav bar 1`] = `
Array [
  <li
    className="moonstone-primaryNavItemsGroup"
  >
    <hr
      className="moonstone-separator moonstone-separator_horizontal moonstone-size_large moonstone-spacing_small"
    />
  </li>,
  <li
    className="moonstone-primaryNavItemsGroup"
  >
    <ul />
  </li>,
]
`;

exports[`primary nav group collapsed and not empty nav bar 1`] = `
Array [
  <li
    className="moonstone-primaryNavItemsGroup"
  >
    <hr
      className="moonstone-separator moonstone-separator_horizontal moonstone-size_large moonstone-spacing_small"
    />
  </li>,
  <li
    className="moonstone-primaryNavItemsGroup"
  >
    <ul />
  </li>,
]
`;

exports[`primary nav group expanded and empty nav bar 1`] = `null`;

exports[`primary nav group expanded and not empty nav bar 1`] = `null`;
